.dotnet/src/Custom/FineTuning/FineTuningOperation.Protocol.cs

```diff
@@ -0,0 +1,430 @@
+        string status = doc.RootElement.GetProperty("status"u8).GetString()!;
+
+        return new FineTuningOperation(client.Pipeline, client.Endpoint, token.JobId, status, response);
+    }
+
+    /// <inheritdoc/>
+    public override async Task WaitForCompletionAsync(CancellationToken cancellationToken = default)
+    {
+        _pollingInterval ??= new();
+
+        while (!IsCompleted)
+        {
+            PipelineResponse response = GetRawResponse();
+
+            await _pollingInterval.WaitAsync(response, cancellationToken);
+
+            ClientResult result = await GetJobAsync(_jobId, cancellationToken.ToRequestOptions()).ConfigureAwait(false);
+
+            ApplyUpdate(result);
+        }
+    }
+
+    /// <summary>
+    /// Waits for the operation to complete processing on the service.
+    /// </summary>
+    /// <param name="pollingInterval"> The time to wait between sending requests
+    /// for status updates from the service. </param>
+    /// <param name="cancellationToken"> A token that can be used to cancel this
+    /// method call. </param>
+    public async Task WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken = default)
+    {
+        _pollingInterval = new(pollingInterval);
+
+        await WaitForCompletionAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public override void WaitForCompletion(CancellationToken cancellationToken = default)
+    {
+        _pollingInterval ??= new();
+
+        while (!IsCompleted)
+        {
+            PipelineResponse response = GetRawResponse();
+
+            _pollingInterval.Wait(response, cancellationToken);
+
+            ClientResult result = GetJob(_jobId, cancellationToken.ToRequestOptions());
+
+            ApplyUpdate(result);
+        }
+    }
+
+    /// <summary>
+    /// Waits for the operation to complete processing on the service.
+    /// </summary>
+    /// <param name="pollingInterval"> The time to wait between sending requests
+    /// for status updates from the service. </param>
+    /// <param name="cancellationToken"> A token that can be used to cancel this
+    /// method call. </param>
+    public void WaitForCompletion(TimeSpan pollingInterval, CancellationToken cancellationToken = default)
+    {
+        _pollingInterval = new(pollingInterval);
+
+        WaitForCompletion(cancellationToken);
+    }
+
+    private void ApplyUpdate(ClientResult result)
+    {
+        PipelineResponse response = result.GetRawResponse();
+
+        using JsonDocument doc = JsonDocument.Parse(response.Content);
+        string? status = doc.RootElement.GetProperty("status"u8).GetString();
+
+        IsCompleted = GetIsCompleted(status);
+        SetRawResponse(response);
+    }
+
+    private static bool GetIsCompleted(string? status)
+    {
+        return status == "succeeded" ||
+            status == "failed" ||
+            status == "cancelled";
```

Comment:
Could we use `InternalFineTuningJobStatus`?
🔗 https://github.com/joseharriaga/openai-in-typespec/blob/main/.dotnet/src/Generated/Models/InternalFineTuningJobStatus.cs